packages/strapi-hook-mongoose/lib/index.js

```diff
@@ -88,16 +88,29 @@ module.exports = function(strapi) {
         connectOptions.useCreateIndex = true;
         connectOptions.useUnifiedTopology = useUnifiedTopology || 'false';
 
-        try {
+        let connect = function() {
           /* FIXME: for now, mongoose doesn't support srv auth except the way including user/pass in URI.
            * https://github.com/Automattic/mongoose/issues/6881 */
-          await instance.connect(
-            uri ||
+          return instance.connect(
+              uri ||
               `mongodb${isSrv ? '+srv' : ''}://${username}:${encodeURIComponent(
-                password
+                  password
               )}@${host}${!isSrv ? ':' + port : ''}/`,
-            connectOptions
+              connectOptions
           );
+        }
+
+        try {
+          await connect();
+
+          instance.connection.on('disconnected', function() {
+            setTimeout(function() {
```

Comment:
why are you using a 2sec timeout ?

Reply:
The 'event' disconnected' is emitted immediately after disconnect. If I call it without setTimeout it will take too much CPU time. 2 seconds is enough fast.

Reply:
Why you it take more cpu time to be called just after the event or 2 sec after ? can you explain

Reply:
Let's assume that a connection at the moment is broken and cannot be established.
After 'disconnected' event I call Mongoose.connect. It causes immediate new 'disconnected' event etc.
So this loop will continue until a connection will be established successfully.
Between event and Mongoose.connect there is no delay or timeout to wait. It is endless recursive loop which takes 100% cpu time, which is not a javascript manner.
So, I put setTimeout to run a connect from global context.

Reply:
Yep totally right. makes sense.

How does this play with the already existing autoReconnect feature for the non replica set usecase ?

We should add settings to configure this new manual reconnect. We can use the same values as the default mongoose settings named `reconnectInterval` and `reconnectTries` to control the number of retries.

Reply:
I think without autoreconnect any strapi project is useless. Everything is based on DB connection. Usual user should not know underlying mechanism to keep connection persistent.
It should work first. I believe I dont need to configure something to make it work because it is a main purpose of the framework - to guarantee that a project is alive.

I'm struggling this issue persistent on production with two different projects and waiting for quick resolving. 

Without replica set a disconnect event is threw when retries over. So, my patch resolves standalone or sharding servers as well.

Reply:
Please add the options with nice defaults so the user doesn't have to configure it but can in case he wants

Reply:
Agreed, it would be nice to allow for a config option in the case of the bookshelf also has a similar setting such as the hook timeout where it will wait for a connection, not sure if mongoose has something similar built in, but if it does that method should be used. In either case allowing the user to customize this setting is a good idea

Reply:
Yes. you can use the options object to add those two options

Reply:
Done